### PR TITLE
feat(notification): #CRRE-480 add notification for validator and prescriber

### PR DESCRIPTION
**Pull request by BastienBenard**
## Describe your changes
Addition of notification in the timeline, for the prescriber and the validator.

## Checklist tests
When a validator finishes processing a basket, the prescriber receives one of the following 3 messages
Your basket is sent to the region
Your basket is partially sent to the region
Your basket is refused

When the region admin finishes processing an order, the validators of the order establishment receive one of the following messages:
One of your establishment's orders has been sent to the bookseller.
One of your establishment's orders has been partially sent to the bookseller.
One of your establishment's orders has been validated by the Île-de-France Region.
One of your establishment's orders has been partially validated by the Île-de-France Region.
One of your establishment's orders has been refused by the Île-de-France Region.

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [x] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

 m

### CODE REVIEW

File: src/main/java/fr/openent/crre/service/impl/DefaultNotificationService.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            return;
+        }
+
+        String i18nStatus = this.getI18nStatusToValidator(orderRegionEquipmentModels);
+        UserUtils.getUserInfos(this.eventBus, userId, userInfos -> {
+            String local = null;
+            try {
+                local = new JsonObject((String) ((LinkedHashMap<?, ?>) userInfos.getAttribute(Field.PREFERENCES)).get(Field.LANGUAGE)).getString(Field.DEFAULT_DASH_DOMAIN);
+            } catch (Exception ignored) {
+            }
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/crre/service/impl/DefaultNotificationService.java:115]: Je ne comprend pas ici le faite de laisser un catch vide.

File: src/main/java/fr/openent/crre/service/impl/DefaultNotificationService.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            return;
+        }
+
+        String i18nStatus = this.getI18nStatusToPrescriber(orderClientEquipmentModels);
+        UserUtils.getUserInfos(this.eventBus, basketOrder.getIdUser(), userInfos -> {
+            String local = null;
+            try {
+                local = new JsonObject((String) ((LinkedHashMap<?, ?>) userInfos.getAttribute(Field.PREFERENCES)).get(Field.LANGUAGE)).getString(Field.DEFAULT_DASH_DOMAIN);
+            } catch (Exception ignored) {
+            }
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/crre/service/impl/DefaultNotificationService.java:132]: CF, commentaire au dessus

**Review by qpic-cgi — APPROVED** [review]

File: src/main/java/fr/openent/crre/controllers/OrderRegionController.java [hunk 3]
```diff
@@ -855,179 +859,75 @@ private void getSearchByIdsOld(HttpServerRequest request, List<JsonArray> result
...
-                    break;
-                }
-                case "consumable_credits": {
-                    purseService.updatePurseAmount(price,
-                            newOrder.getString("id_structure"), operation, true,
-                            handlerJsonObject(updateFuture));
-                    break;
-                }
+    private TransactionElement getUpdateTransactionElement(OrderClientEquipmentType status, OrderRegionEquipmentModel orderRegion) {
+        if ("REJECTED".equals(orderRegion.getStatus())) {
```
Comment (Sofianebernoussi) [inline src/main/java/fr/openent/crre/controllers/OrderRegionController.java:920]: t'as créer un enum pour le rejected je crois autant l'utiliser

File: src/main/java/fr/openent/crre/model/neo4j/Neo4jUserDataModel.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package fr.openent.crre.model.neo4j;
+
+import fr.openent.crre.core.constants.Field;
+import fr.openent.crre.helpers.IModelHelper;
+import fr.openent.crre.model.IModel;
+import io.vertx.core.json.JsonObject;
+
+public class Neo4jUserDataModel implements IModel<Neo4jUserDataModel> {
```
Comment (Sofianebernoussi) [inline src/main/java/fr/openent/crre/model/neo4j/Neo4jUserDataModel.java:8]: la classe userinfo peut être utiliser a la lmite l'extend

File: src/main/java/fr/openent/crre/model/neo4j/Neo4jUserModel.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package fr.openent.crre.model.neo4j;
+
+import fr.openent.crre.core.constants.Field;
+import fr.openent.crre.helpers.IModelHelper;
+import fr.openent.crre.model.IModel;
+import io.vertx.core.json.JsonObject;
+
+public class Neo4jUserModel implements IModel<Neo4jUserModel> {
```
Comment (Sofianebernoussi) [inline src/main/java/fr/openent/crre/model/neo4j/Neo4jUserModel.java:8]: pareil

**Review by sjollois — APPROVED** [review]

File: src/main/resources/view-src/notify/new_notification.html [hunk 6]
```diff
@@ -0,0 +1,17 @@
+
+<div class="crre-timeline-new-notification">
```
Comment (JinDouang) [inline src/main/resources/view-src/notify/new_notification.html:2]: Je te fais un commentaire côté css : https://github.com/CGI-OPEN-ENT-NG/entcore-css-lib/pull/32/files#diff-14afd4688fa684781c4f1e209b9e21a1e00e646f65f50b07c82fcf9a76693f78R1

File: src/main/java/fr/openent/crre/service/impl/DefaultNotificationService.java (hunk 1 above)
Comment (BastienBenard) [reply src/main/java/fr/openent/crre/service/impl/DefaultNotificationService.java:115]: S'il y a une erreur je n'ai pas besoin de faire autre chose. Je peux faire ça mais cela reviens au meme
```suggestion
            String local;
            try {
                local = new JsonObject((String) ((LinkedHashMap<?, ?>) userInfos.getAttribute(Field.PREFERENCES)).get(Field.LANGUAGE)).getString(Field.DEFAULT_DASH_DOMAIN);
            } catch (Exception ignored) {
                local = null;
            }
```

Comment (qpic-cgi) [reply src/main/java/fr/openent/crre/service/impl/DefaultNotificationService.java:115]: tu ne peux pas enlever le try du coup ?

Comment (BastienBenard) [reply src/main/java/fr/openent/crre/service/impl/DefaultNotificationService.java:115]: non car s'il y a un exception telque un CastException, oon ne va pas envoyer la notification

File: src/main/resources/view-src/notify/new_notification.html (hunk 6 above)
Comment (BastienBenard) [reply src/main/resources/view-src/notify/new_notification.html]: Vu avec Jin et Laurie, ce n'est plus d'actualiter. On ne veut pas de logo custom.

**Review by JinDouang — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/crre/service/impl/DefaultUserService.java [hunk 7]
```diff
@@ -18,4 +31,34 @@ public void getStructures(String userId, Handler<Either<String, JsonArray>> hand
...
+    @Override
+    public Future<Map<Neo4jUserModel, String>> getValidatorUser(List<String> structureIdList) {
+        Promise<Map<Neo4jUserModel, String>> promise = Promise.promise();
+
+        String query = "MATCH (s:Structure)<--()--(u:User)-->(g:Group)-->(r:Role)-[:AUTHORIZE]->(w:WorkflowAction{displayName:'" + Crre.VALIDATOR_RIGHT + "'})" +
+                " WHERE s.id IN {structureIdList}" +
+                " WITH r,u,s" +
+                " MATCH (wa:WorkflowAction{displayName:'" + Crre.ADMINISTRATOR_RIGHT + "'})" +
+                " WHERE NOT ((r)-[:AUTHORIZE]->(wa))" +
+                " return distinct u,s.id";
```
[the remark this answers is not in the document]
Comment (BastienBenard) [reply src/main/java/fr/openent/crre/service/impl/DefaultUserService.java:44]: Le data ne sert a rien dans Neo4jUserModel, je vais le supprimer. Pour le coups je laisse u dans le cas on l'on aurait besoin de plus d'information pour les futures dev

**Review by Sofianebernoussi — APPROVED** [review]

**Review by JinDouang — APPROVED** [review]

**Pull request merged**